fix(ssr): global attribute disparities

jhefferman-sfdc opened the pull request:
## Details

## Does this pull request introduce a breaking change?

- 😮‍💨 No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

- 🤞 No, it does not introduce an observable change.

## GUS work item

[W-17572462](https://gus.lightning.force.com/a07EE000027ljyNYAQ)

jhefferman-sfdc commented on packages/@lwc/engine-server/src/__tests__/fixtures/attribute-global-html/as-component-prop/undeclared/modules/x/component/component.html:
```diff
@@ -1,4 +1,11 @@
 <template>
+    <!--
```
These 4 deviate between v1/2, where v2 aligns with CSR and we don't want to fix V1

jhefferman-sfdc commented on packages/@lwc/engine-server/src/__tests__/fixtures/attribute-global-html/as-component-prop/without-@api-values/config.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "entry": "x/component",
```
V1/V2 deviate, where V2 parent overrides the child value (as does CSR). V1 is incorrect.

jhefferman-sfdc commented on packages/@lwc/engine-server/src/__tests__/fixtures/attribute-global-html/as-component-prop/without-@api/config.json:
```diff
@@ -1,3 +1,7 @@
-    "entry": "x/component"
+    "entry": "x/component",
+    "ssrFiles": {
+        "error": "error-ssr.txt",
```
V1/V2 deviate, where V2 parent overrides the child value (as does CSR). V1 is incorrect.

jhefferman-sfdc commented on packages/@lwc/engine-server/src/__tests__/fixtures/attribute-global-html/draggable/config.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "entry": "x/component",
+    "ssrFiles": {
+        "error": "error-ssr.txt",
```
V1/V2 deviate, where V2 matches CSR

jhefferman-sfdc commented on packages/@lwc/engine-server/src/__tests__/fixtures/attribute-global-html/hidden/config.json:
```diff
@@ -0,0 +1,8 @@
+    
+    "entry": "x/component",
+    "ssrFiles": {
+        "error": "error-ssr.txt",
```
V1/V2 deviate, where V2 matches CSR

jhefferman-sfdc commented on packages/@lwc/engine-server/src/__tests__/fixtures/attribute-global-html/spellcheck/config.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "entry": "x/component",
+    "ssrFiles": {
```
V1/V2 deviate, where V2 matches CSR

jhefferman-sfdc commented on packages/@lwc/engine-server/src/__tests__/fixtures/attribute-global-html/tabindex/config.json:
```diff
@@ -0,0 +1,7 @@
+    "entry": "x/component",
+    "ssrFiles": {
+        "error": "error-ssr.txt",
+        "expected": "expected-ssr.html"
```
V1/V2 deviate, where V2 matches CSR

jhefferman-sfdc commented on packages/@lwc/engine-server/src/__tests__/fixtures/known-boolean-attributes/default-def-html-attributes/static-on-component/config.json:
```diff
@@ -1,3 +1,7 @@
-    "entry": "x/component"
+    "entry": "x/component",
+    "ssrFiles": {
+        "error": "error-ssr.txt",
```
V1/V2 deviate on draggable, where V2 matches CSR. Left this in but there are also dedicated tests above

jhefferman-sfdc commented on packages/@lwc/ssr-runtime/src/lightning-element.ts:
```diff
@@ -92,13 +87,12 @@ export class LightningElement implements PropsAvailableAtConstruction {
         // Avoid setting the following types of properties that should not be set:
         // - Properties that are not public.
         // - Properties that are not global.
-        // - Properties that are global but are internally overridden.
```
Does not seem correct. In CSR, child definitions do not override parent.

wjhsf replied:
If we had fixture tests for this, and it wasn't caught, we should add some karma tests to get the real browser behavior.

jhefferman-sfdc replied:
yeah i'm adding some karma for these

wjhsf commented on packages/@lwc/ssr-compiler/src/compile-template/shared.ts:
```diff
@@ -153,20 +154,27 @@ export function getChildAttrsOrProps(
+                    literalValue = shouldNormalize ? '0' : literalValue;
                 }
                 return b.property('init', key, b.literal(literalValue));
             } else if (value.type === 'Literal' && typeof value.value === 'boolean') {
```
What happens if `tabindex` is set to a boolean value? Can it be?

jhefferman-sfdc replied:
Yup, per the strange CSR logic, [booleans are preserved](https://github.com/salesforce/lwc/blob/f34a347/packages/%40lwc/engine-core/src/framework/api.ts#L193-L211), so nothing to do for this case here.

wjhsf commented on packages/@lwc/ssr-compiler/src/compile-js/generate-markup.ts:
```diff
@@ -18,7 +18,6 @@ const bGenerateMarkup = esTemplate`
     // These variables may mix with component-authored variables, so should be reasonably unique
     const __lwcSuperPublicProperties__ = Array.from(Object.getPrototypeOf(${/* Component class */ is.identifier})?.__lwcPublicProperties__?.values?.() ?? []);
     const __lwcPublicProperties__ = new Set(${/*public properties*/ is.arrayExpression}.concat(__lwcSuperPublicProperties__));
-    const __lwcPrivateProperties__ = new Set(${/*private properties*/ is.arrayExpression});
```
Deleting stuff! 🚀

jhefferman-sfdc commented:
/nucleus ignore --reason "Unrelated __lwc failure in webruntime"

wjhsf commented on packages/@lwc/ssr-compiler/src/compile-template/shared.ts:
```diff
@@ -153,20 +154,27 @@ export function getChildAttrsOrProps(
                     // https://github.com/salesforce/lwc/blob/574ffbd/packages/%40lwc/template-compiler/src/codegen/index.ts#L445-L448
                     literalValue = literalValue.toLowerCase() !== 'false';
+                } else if (name.toLowerCase() === 'tabindex') {
+                    // Global HTML "tabindex" attribute is specially massaged into a stringified number
+                    // This follows the historical behavior in api.ts:
+                    // https://github.com/salesforce/lwc/blob/f34a347/packages/%40lwc/engine-core/src/framework/api.ts#L193-L211
+                    const shouldNormalize = parseInt(literalValue) > 0;
+                    literalValue = shouldNormalize ? '0' : literalValue;
```
```suggestion
                    literalValue = normalizeTabIndex(literalValue);
```
If we've got the util, use it!

wjhsf commented on packages/@lwc/shared/src/normalize-tab-index.ts:
```diff
@@ -0,0 +1,16 @@
+ * This follows the historical behavior in api.ts:
+ * https://github.com/salesforce/lwc/blob/f34a347/packages/%40lwc/engine-core/src/framework/api.ts#L193-L211
+ */
+export function normalizeTabIndex(value: any): any {
```
Add some unit tests for this!

wjhsf commented on packages/@lwc/shared/src/normalize-tab-index.ts:
```diff
@@ -0,0 +1,16 @@
+ * This follows the historical behavior in api.ts:
+ * https://github.com/salesforce/lwc/blob/f34a347/packages/%40lwc/engine-core/src/framework/api.ts#L193-L211
+ */
+export function normalizeTabIndex(value: any): any {
```
```suggestion
export function normalizeTabIndex(value: number): string | number {
```
The original function was more broadly typed than necessary, since this is new we can be more restrictive.

Also, it's weird that this returns either the unmodified number or the string `'0'`. We should either always return a number or always return a string. (It's diverging from the original implementation, but not in a meaningful way.)

wjhsf replied:
In the end, the attribute is going to end up a string attribute / numeric prop. This particular function is just an intermediate step in that. This particular function copies the business logic from `engine-core`, but the rest of the implementation (outside the function) is different, so I don't think it makes much of a difference whether the final cast to string/number happens in or out of the function.

wjhsf replied:
```suggestion
export function normalizeTabIndex<T>(value: T): T | string {
```
If you want to allow any type.

jhefferman-sfdc replied:
I tried generic but it added issues with the comparison. I wanted to keep it exactly as the original to avoid any discrepancies

wjhsf commented on packages/@lwc/engine-server/src/__tests__/fixtures/attribute-global-html/tabindex/modules/x/component/component.js:
```diff
@@ -0,0 +1,12 @@
+
+export default class extends LightningElement {
+    undefined = undefined;
+    aNull = null;
```
🤨

jhefferman-sfdc replied:
ahahaha

jhefferman-sfdc replied:
i just got this

divmain commented on packages/@lwc/ssr-compiler/src/compile-template/shared.ts:
```diff
@@ -153,20 +154,27 @@ export function getChildAttrsOrProps(
                 if (name === 'class') {
                     cxt.import('normalizeClass');
                     propValue = b.callExpression(b.identifier('normalizeClass'), [propValue]);
+                } else if (name.toLowerCase() === 'tabindex') {
```
nit: there is more than one `name.toLowerCase()` showing up in this function; probably could define `nameLower` variable.

jhefferman-sfdc commented:
/nucleus ignore --reason "Unrelated __lwc failure in webruntime"

jhefferman-sfdc commented:
I added hydration coverage for global property precedence. As SSR V1 renders the wrong values, it causes hydration errors. I added a check [here](https://github.com/salesforce/lwc/pull/5247/files#diff-4f73a49efd56c88d755e7a1e60dfd353eef5f9cfb8d54e4a75e312234e796b14R29) and an expectation for those failures. Here is the work item to review this and change the pattern if needed: W-17939029
